packages/engine/Source/Renderer/TextureAtlas.js

```diff
@@ -671,54 +679,76 @@ TextureAtlas.prototype.addImage = function (id, image) {
       return -1;
     }
 
-    return this._addImage(index, image);
+    return this._addImage(index, image).then((index) => {
+      this._indexPromiseById.delete(id);
+      return index;
+    });
   };
 
   promise = resolveAndAddImage();
   this._indexPromiseById.set(id, promise);
   return promise;
 };
 
+/**
+ * Get an existing sub-region of an existing atlas image as additional image indices.
+ * @private
+ * @param {string} id The identifier of the existing image.
+ * @param {BoundingRectangle} subRegion An {@link BoundingRectangle} defining a region of an existing image, measured in pixels from the bottom-left of the image.
+ * @param {number} imageIndex The index of the image.
+ * @returns {Promise<number> | number | undefined} The existing subRegion index, or undefined if not yet added.
+ */
+TextureAtlas.prototype.getCachedImageSubRegion = function (
+  id,
+  subRegion,
+  imageIndex,
+) {
+  const imagePromise = this._indexPromiseById.get(id);
+  for (const [index, parentIndex] of this._subRegions.entries()) {
+    if (imageIndex === parentIndex) {
+      const boundingRegion = this._rectangles[index];
+      if (boundingRegion.equals(subRegion)) {
+        // The subregion is already being tracked
+        if (imagePromise) {
+          return imagePromise.then((resolvedImageIndex) =>
```

Comment:
Unlike the previous commit, where the `subRegion` could often be cached and **synchronously** return an index before the image was fully loaded, here I make sure that the `subRegion` can never return a synchronous value until the underlying image is fully loaded and resolved.